app/util/event_log.py

```diff
@@ -0,0 +1,265 @@
+
+    def _write_to_log_cache(self, event_data):
+        """
+        :param event_data: Free-form key value pairs that make up the event
+        :type event_data: dict
+        """
+        self._log_cache.append(event_data)
+        self._expire_stale_items_in_cache()
+
+    def _expire_stale_items_in_cache(self):
+        """
+        Expires stale items in the cache. An item in the cache is considered stale if the timestamp is
+        significantly older than the most recent item. This expiration time is determined by the constant,
+        LOG_CACHE_EXPIRE_TIME_IN_HOURS.
+
+        Since we do not want to artificially impose limits on the log cache, we will not expire items
+        if the total length of the log cache is less than LOG_CACHE_THRESHOLD.
+        """
+        while len(self._log_cache) > LOG_CACHE_THRESHOLD and self.oldest_cache_event_is_stale():
+            self._log_cache.popleft()
+
+    def oldest_cache_event_is_stale(self):
+        """
+        Returns whether the time stamp of the event data is stale in comparison with the start time.
+        The time is considered stale if it is is older than start_time + LOG_CACHE_EXPIRE_TIME_IN_HOURS
+
+        :rtype: bool
+        """
+        if len(self._log_cache) == 0:
+            return False
+        else:
+            latest = self._latest_timestamp_in_cache()
+            oldest = self._oldest_timestamp_in_cache()
+            return oldest + LOG_CACHE_EXPIRE_TIME_IN_HOURS * 60 * 60 < latest
+
+    def _latest_timestamp_in_cache(self):
+        """
+        :rtype: float | None
+        """
+        if len(self._log_cache) == 0:
+            return None
+        else:
+            return self._log_cache[-1]['__timestamp__']
+
+    def _oldest_timestamp_in_cache(self):
+        """
+        :rtype: float | None
+        """
+        if len(self._log_cache) == 0:
+            return None
+        else:
+            return self._log_cache[0]['__timestamp__']
+
+    def get_events(self, since_timestamp=None, since_id=None):
+        """
+        Retrieve all events from the current eventlog since the given timestamp or event id.
+        This is used to expose events via the API and is useful for building dashboards
+        that monitor the system.
+
+        :param since_timestamp: Get all events after (greater than) this timestamp
+        :type since_timestamp: float | None
+        :param since_id:  Get all events after (greater than) this id.
+        :type since_id: int | None
+        :return: The list of events in the given range
+        :rtype: list[dict] | None
+        """
+        if self.logging_disabled or self.filename.upper() == 'STDOUT':
+            return None
+
+        if since_timestamp is not None and since_id is not None:
+            raise ValueError('since_timestamp and since_id can not be used at the same time')
+        events = []
+        condition_met = False
```

Comment:
As soon as I read the name `condition_met`, my mind immediately goes "What condition?!" :bowtie: 

What do you think of something like `is_query_complete`?

---

As I look at this more, it seems like the `condition_met` flag might not be doing much? It looks like the flow is:


Alternatively, we could just do the condition check in this method after querying the cache instead of having the `_get_events_from_reversed_generator` method do it. Something like:
```
if not self._query_extends_beyond_cache(since_timestamp):
    events = self._get_event_from_cache(since_timestamp)


```